[jax2tf(js)] Convertible QuickDraw model using Linen

bchetioui opened the pull request:
This builds on #4522, #4562 and #4589 to allow converting a Flax model into TF.js for inference.
Just opening as draft for now, let's see what exactly we do with this.

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+
+TfVariable = Any
+
+class ExportWrapper(tf.train.Checkpoint):
```
I would try to not make changes to the ExportWrapper because we expect to make that a library function that just implements: https://www.tensorflow.org/hub/reusable_saved_models

gnecula replied:
Also, I would like to know what is different here.

bchetioui replied:
I undid the differences I made. The only difference that is important is that `with_gradient` must be set to True, so as to avoid having `PreventGradient` ops in the graph (as they are not supported by Tensorflow JS)

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+    """
+    super().__init__()
+    # Convert fn from JAX to TF.
+    self._fn = jax2tf.convert(fn, with_gradient=True)
```
Was it important to add `with_gradient`?

bchetioui replied:
Yes, as `PreventGradient` is not a valid op for the conversion!

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+    # Create tf.Variables for the parameters.
+    self._params = tf.nest.map_structure(
+        lambda param: tf.Variable(param, trainable=False), params)
+    self.saved_signatures = {}
```
Afaik, `saved_signatures` is not part of the API from `reusable_saved_models`. Did you add it just as a place to store some data for your need, or it actually matters to be passed to `tf.saved_model.save`?

bchetioui replied:
It is solely to pass them to `tf.saved_model.save`, but they can be passed manually as you saw below. So it is perfectly fine to store them outside of the `ExportWrapper`.

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+    # to work properly, but we might be able to change that through some
+    # parameter of the converter.
+    self.saved_signatures["serving_default"] = (
+      self.__call__.get_concrete_function(
```
Isn't this achieved also by the code in line 73?

bchetioui replied:
It is! Only that you then need to explicitly pass a `signature_def` argument to `convert_tf_saved_model` to tell which signature to use, as it looks for one called `serving_default` otherwise

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+    fn, params, input_shape,
+    batch_sizes=batch_sizes)
+  print(f"Saving the model to {model_dir}")
+  tf.saved_model.save(wrapper, model_dir, signatures=wrapper.saved_signatures)
```
If this is the only place where we use `saved_signatures` then I think it is better to not have them in the ExportWrapper.

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+      print("Test set accuracy {}".format(test_acc))
+
+    return (FlaxQuickDraw.predict, optimizer.target,
+            FlaxQuickDraw.feature_extractor,
```
We don't need the feature extractor and its params, let's inline the feature extractor in the model.

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+# The code below is an adaptation for Flax from the work published here:
+# https://blog.tensorflow.org/2018/07/train-model-in-tfkeras-with-colab-and-run-in-browser-tensorflowjs.html
+
+class FlaxQuickDraw:
```
This class existed in the other notebook because I had two experiments that I wanted to keep separate. I think that here we can just move everything to the top-level

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+
+    @nn.compact
+    def __call__(self, x):
+      x = self.feature_extractor(x)
```
Just inline here the feature extractor module

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+      test_acc = FlaxQuickDraw.accuracy(
+          FlaxQuickDraw.predict, optimizer.target, test_ds)
+      print("Epoch {} in {:0.2f} sec".format(epoch, epoch_time))
+      print("Training set accuracy {}".format(train_acc))
```
To what accuracy were you able to train it? and how long it took?

bchetioui replied:
Took roughly 27 min to train it with 77% accuracy (where we consider a prediction to be accurate if the expected label is the most likely prediction, which is different from the original TF example which considers a prediction accurate if the expected label is one of the 5 most likely).

gnecula commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/sketcher.py:
```diff
@@ -0,0 +1,278 @@
+            FlaxQuickDraw.feature_extractor,
+            optimizer.target["feature_extractor"])
+
+def download_dataset(dir_path=os.path.join(os.path.dirname(__file__), "data")):
```
This has to be called before? How long does the download take and how big is the data set?

bchetioui replied:
It depends on how many classes we want to use. The example had 100, with each data file roughly ~100 to 200 Mb (so 10 to 20 Gb total if we stick to this number of classes)

bchetioui replied:
After getting the whole dataset, I can say: the download is not *that* long (of course, it depends on your connection speed), and the considered subset of 100 classes is roughly 11 Gb. How many classes we want to work with (and thus how many files we want to download) can be adjusted using the `NB_CLASSES` global variable in `sketcher.py`.

bchetioui commented on jax/experimental/jax2tf/tfjs_demo/Sketcher/utils.py:
```diff
@@ -0,0 +1,152 @@
+        # (even with hub.KerasLayer(..., trainable=False), surprisingly).
+        lambda param: tf.Variable(param, trainable=with_gradient),
+        params)
+    self._signatures: Dict[str, Callable] = {}
+
```
@gnecula this was part of the `ExportWrapper`, but I don't see it mentioned in the spec. Is it needed?

gnecula replied:
I do not think so.

bchetioui replied:
Thank you, I removed it.

bchetioui committed:
[jax2tf(js)] Implement a CNN based on the Quickdraw dataset that
is convertible to TF.js.

- the demo is based on https://blog.tensorflow.org/2018/07/train-model-in-tfkeras-with-colab-and-run-in-browser-tensorflowjs.html
- the third_party code we include in the repo is the one used in
that demo and hosted at https://github.com/zaidalyafeai/zaidalyafeai.github.io/tree/master/sketcher

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/quickdraw.py:
```diff
@@ -0,0 +1,147 @@
+  classes = download_dataset(dir_path, NB_CLASSES)
+  assert len(classes) == NB_CLASSES, classes
+  print(f"Classes are: {classes}")
+  flax_params = train(dir_path, classes)
```
nit: What do you think about loading the classes here, and calling `train` with the data sets? Then all of the interaction with `dir_path` is in the `main` function and we just pass data around in the other functions.

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/quickdraw.py:
```diff
@@ -0,0 +1,147 @@
+  print(f"Classes are: {classes}")
+  flax_params = train(dir_path, classes)
+
+  model_dir = os.path.join(os.path.dirname(__file__), "saved_models")
```
I think that it would be better to use a temporary directory for saving the models. Saving next to the sources will never work when run inside Google. I suggest either hard-coding /tmp/jax2tf/tf_js_quickdraw or `tempfile`.

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/README.md:
```diff
@@ -0,0 +1,8 @@
+This example contains the code necessary to train a CNN on the Quickdraw
+dataset using Flax, and subsequently convert it using JAX2TF into a model that
+can be loaded in TF.js.
+
```
Can you please add some text about what the user is supposed to run, and what to look for? Also perhaps what to do with the tarball and how it fits with the rest of the code here?

bchetioui replied:
PTAL at the latest README and let me know what you think!

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/utils.py:
```diff
@@ -0,0 +1,75 @@
+def download_dataset(dir_path, nb_classes):
+
+  if not os.path.exists(dir_path):
+    os.mkdir(dir_path)
```
I think that `os.makedirs` will create the directory and its parents, if needed.

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/quickdraw.py:
```diff
@@ -0,0 +1,147 @@
+from jax.experimental import jax2tf # type: ignore
+jax2tf.jax2tf.ENABLE_TF_CONVOLUTION = True
+
+from jax.experimental.jax2tf.examples.saved_model_lib import save_model # type: ignore
```
I suggest importing modules, not functions:
```
from jax.experimental.jax2tf.examples impot saved_model_lib
```

Then when we see `saved_model_lib.save_model` we know right away what is being called.

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/quickdraw.py:
```diff
@@ -0,0 +1,149 @@
+  return optimizer.target
+
+def main(*args):
+  dir_path = os.path.join(os.path.dirname(__file__), "data")
```
Please change this one to base_model_path/data

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/quickdraw.py:
```diff
@@ -0,0 +1,149 @@
+  # the model must be converted with with_gradient set to True to be able to
+  # convert the saved model to TF.js, as "PreventGradient" is not supported
+  input_signatures = [
+    tf.TensorSpec([batch_size, 28, 28, 1]) for batch_size in [1, 256]
```
Do we even need to save for batch_size 256? Do you know how convert_tf_saved_model handles the batch size? It converts only the first input_signature? Converts all (I'd be surprised)?

bchetioui replied:
No, it is indeed useless. It converts the one corresponding to the `serving_default` signature, which I believe would now be the first signature with the new Exporter code you pushed this morning

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/README.md:
```diff
@@ -0,0 +1,38 @@
+Here is a (third party) [demo](https://zaidalyafeai.github.io/sketcher/) of how
+interacting with the trained model looks like.
+
+If you want to try it yourself, the trained model along with third party front
```
You said that there is a readme in the tarball. Perhaps it is worth saying here also that the tarball has the same exact files as the above linked website, with one line change to ..., along with the trained weights. Do you happen to have training accuracy numbers?

bchetioui replied:
I will upgrade. Yes, I have training accuracy numbers:
```
Training set accuracy 0.7825493812561035
Test set accuracy 0.764135718345642
```
where "accurate" means "the prediction with the highest likelihood is the expected class".

gnecula replied:
Please update the accuracy numbers in the README

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/quickdraw.py:
```diff
@@ -0,0 +1,146 @@
+from jax import lax
+from jax import numpy as jnp
+from jax.experimental import jax2tf # type: ignore
+jax2tf.jax2tf.ENABLE_TF_CONVOLUTION = True
```
Let's wait until you land the other mechanism.

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/quickdraw.py:
```diff
@@ -0,0 +1,146 @@
+  model_dir = os.path.join(base_model_path, "saved_models")
+  # the model must be converted with with_gradient set to True to be able to
+  # convert the saved model to TF.js, as "PreventGradient" is not supported
+  saved_model_lib.save_model(predict, flax_params, model_dir,
```
Have you tried this with the now default "convert_model=True"?

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Training the model
+
+To train and export the model yourself, simply run
```
Is this step going to take a while? It would be good to tell the user, something like "You can train and export the model yourself, which will take 30 minutes, or you can skip to [next section](#interacting-with-the-model)"

gnecula commented on jax/experimental/jax2tf/examples/README.md:
```diff
@@ -13,6 +13,10 @@ At the moment, the open-source TensorFlow model server is missing XLA support,
 
+A jax2tf-generated SavedModel can also be converted to a format usable with
+TensorFlow.js in some cases where the conversion does not require XLA support,
+as show in the [Quickdraw TensorFlow.js example](https://github.com/google/jax/blob/master/jax/experimental/jax2tf/examples/tf_js/quickdraw/README.md).
```
as shown

gnecula commented on jax/experimental/jax2tf/examples/tf_js/quickdraw/README.md:
````diff
@@ -0,0 +1,48 @@
+```
+
+The training itself lasts roughly 30 minutes, assuming the dataset has already
+been downloaded. You can also skip to the [next section](https://github.com/google/jax/blob/master/jax/experimental/jax2tf/examples/tf_js/quickdraw/README.md#interacting-with-the-model)
````
nit: In this particular case you can just write [#interacting-with-the-model] because that works both in github and in Google.
